### PR TITLE
Unbreak ghcide and bump haskell-lsp version

### PR DESCRIPTION
**Pull request by turion**
###### Things done

- [ ] Tested using sandboxing ([nix.useSandbox](http://nixos.org/nixos/manual/options.html#opt-nix.useSandbox) on NixOS, or option `sandbox` in [`nix.conf`](http://nixos.org/nix/manual/#sec-conf-file) on non-NixOS linux)
- Built on platform(s)
   - [ ] NixOS
   - [ ] macOS
   - [ ] other Linux distributions
- [ ] Tested via one or more NixOS test(s) if existing and applicable for the change (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
- [ ] Tested compilation of all pkgs that depend on this change using `nix-shell -p nix-review --run "nix-review wip"`
- [ ] Tested execution of all binary files (usually in `./result/bin/`)
- [ ] Determined the impact on package closure size (by running `nix path-info -S` before and after)
- [ ] Ensured that relevant documentation is up to date
- [ ] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/.github/CONTRIBUTING.md).

###### Notify maintainers

cc @

### CODE REVIEW

File: pkgs/development/haskell-modules/configuration-hackage2nix.yaml
Comment (cdepillabout) [inline pkgs/development/haskell-modules/configuration-hackage2nix.yaml]: In general, the Haskell stuff in nixpkgs tracks the latest LTS release.

If you look at the top of this list here, you can see that these packages are locked to versions from the latest LTS.  (peti updates these weekly, you can watch live if you're interested: https://www.twitch.tv/peti343)

It appears that the latest version of ghcide requires `haskell-lsp ==0.17.*, haskell-lsp-types ==0.17.*`, which is later than the version from LTS-14.

However, hackage2nix generates the latest versions of packages in the LTS as well, so `haskellPackages.haskell-lsp_0_18_0_0` and `haskellPackages.haskell-lsp-types_0_18_0_0` should be available.

I suggest trying to override ghcide to use these two package versions and then jailbreaking the cabal file.

Here's an example of `hnix` overriding package versions as well as using `doJailbreak`:

https://github.com/NixOS/nixpkgs/blob/63f66812bd263f96ec3c335b4160d8b2a2fe69ad/pkgs/development/haskell-modules/configuration-common.nix#L230-L233

Could you update this PR to remove the above lines from `configuration-hackage2nix.yaml`, and instead adding an entry for `ghcide` to `configuration-common.nix`?

**Review by cdepillabout — CHANGES_REQUESTED** [review]

File: pkgs/development/haskell-modules/configuration-hackage2nix.yaml
Comment (cdepillabout) [reply pkgs/development/haskell-modules/configuration-hackage2nix.yaml]: > I suggest trying to override ghcide to use these two package versions and then jailbreaking the cabal file.

Oh, and let me know if this doesn't work.  There is a way to force `hackage2nix` to generate arbitrary package versions (so we could make it generate `haskellPackages.haskell-lsp_0_17_0_0` for `ghcide`).

Comment (turion) [reply pkgs/development/haskell-modules/configuration-hackage2nix.yaml]: Thanks for your explanations! Do I have to run `hackage2nix` locally to test my changes?

Comment (cdepillabout) [reply pkgs/development/haskell-modules/configuration-hackage2nix.yaml]: @turion No, peti runs `hackage2nix` locally, so you don't have to do it in this PR.

Comment (turion) [reply pkgs/development/haskell-modules/configuration-hackage2nix.yaml]: Unfortunately `ghcide` does not build with `haskell-lsp-0.18`. We probably need to make a `haskellPackages.haskell-lsp_0_17_0_0 `, but I don't know how. In parallel, I'll start and upgrade `ghcide` to build with a newer `haskell-lsp`.

Comment (turion) [reply pkgs/development/haskell-modules/configuration-hackage2nix.yaml]: (This is known: https://github.com/digital-asset/ghcide/issues/219 It is fixed on `master`, but there has been no release since.)

Comment (cdepillabout) [reply pkgs/development/haskell-modules/configuration-hackage2nix.yaml]: @turion thanks for the update.  I think there are two possibilities:

1. Get ghcide to release a new version that supports `haskell-lsp-0.18`, and then do as suggested above and send a PR to nixpkgs to override the version of `haskell-lsp` that the `ghcide` package uses.

2. Send a PR to nixpkgs that generates an older version of `haskell-lsp`.  You can do it by adding a line like https://github.com/NixOS/nixpkgs/blob/d0e512d6930490337a51f245f1827fa16ca01528/pkgs/development/haskell-modules/configuration-hackage2nix.yaml#L2510 for `haskell-lsp == 0.17.0.0`

    When this gets merged in (and `hackage-packages.nix` gets regenerated), there will be a new `haskellPackages.haskell-lsp_0_17_0_0` generated.  You can then use this in an override for `ghcide` as described above.

Solution 1 would be easier for us (and better for the community).  However, it is more work for you, and we have to rely on the `ghcide` maintainers merging your PR quickly and making a new release.  If the `ghcide` maintainers aren't quick with responding to your PR and you get tired of waiting, let's go with solution 2.

Comment (turion) [reply pkgs/development/haskell-modules/configuration-hackage2nix.yaml]: I see. I'm working towards 1. then first :)

Comment (turion) [reply pkgs/development/haskell-modules/configuration-hackage2nix.yaml]: `ghcide` with `haskell-lsp-0.18` is released. I don't really understand how this change gets here. Do I just wait and rebase?

Comment (cdepillabout) [reply pkgs/development/haskell-modules/configuration-hackage2nix.yaml]: @turion

> I don't really understand how this change gets here.

I think there is an automatic script that updates the `haskell-updates` branch with recent package versions from Hackage.  (Or maybe it is semi-automatic and peti has to run it by hand)

You can see the current state of `haskell-updates` here (including a bunch of commits with the message `hackage-packages.nix: automatic Haskell package set update`):

https://github.com/NixOS/nixpkgs/pull/75429

At some point (maybe today or tomorrow?) one of these Haskell package set updates should contain the new version of `ghcide`.  At that point you should be able to rebase this PR on `haskell-updates` and use the new version of `haskell-lsp`.

Comment (turion) [reply pkgs/development/haskell-modules/configuration-hackage2nix.yaml]: Ok. Thanks for all your explanations!

Is this whole workflow documented somewhere? Just in case peti goes on a 2 month vacation and someone needs to pick up the reign ;)

Comment (cdepillabout) [reply pkgs/development/haskell-modules/configuration-hackage2nix.yaml]: It is somewhat documented in the `hackage2nix` Readme:

https://github.com/NixOS/cabal2nix/tree/master/hackage2nix

There is also a youtube video showing the steps:

https://www.youtube.com/watch?v=qX0mgtSm360

Also, peti streams live these updates, so if you watch him doing it once or twice, you should get a good idea of how to do it:

https://discourse.nixos.org/t/please-join-the-new-weekly-haskell-updates-video-livestream/4420

**Review by peti — CHANGES_REQUESTED** [review]
Can you please re-base this PR on top of a current version of `haskell-updates` to resolve the merge conflicts?

**Comment by turion** [comment]
> Can you please re-base this PR on top of a current version of `haskell-updates` to resolve the merge conflicts?

@peti Yes, since the conflicts are only internal to `haskell-updates`. But it seems that you added a commit that obsoletes my PR. Also, it seems that it doesn't work anymore because `haskell-lsp` has advanced to 0.19.0.0 in the meantime.

**Comment by turion** [comment]
Another dependency, `hie-bios`, is broken on `haskell-updates`. Something with the test suite. I'll see whether I can fix it easily.

**Comment by turion** [comment]
> Another dependency, `hie-bios`, is broken on `haskell-updates`. Something with the test suite. I'll see whether I can fix it easily.

Seems nontrivial to fix. I'll add `dontCheck` for the time being. This should be removed once there is an acceptable answer to https://github.com/mpickering/hie-bios/issues/25#issuecomment-567011935

**Comment by turion** [comment]
Fixing `hie-bios` is separately in https://github.com/NixOS/nixpkgs/pull/75869.

**Comment by turion** [comment]
I'm a bit out of ideas. The upstream version for `haskell-lsp` is 0.19 by now, and `ghcide` is once again behind. It does not compile with the newer `haskell-lsp`, that's the latest thing I tried now. Do I have to pin the version of `haskell-lsp` at 0.18 somewhere? Is this a manual process that gets stuck every time upstream versions get out of sync?

**Comment by mpickering** [comment]
@turion You are discovering why people use `haskell.nix`.

**Comment by cdepillabout** [comment]
@turion What you can do is send a PR adding a line to this list:

https://github.com/NixOS/nixpkgs/blob/c8c7398d849f5ea844977faf100a69817047daa6/pkgs/development/haskell-modules/configuration-hackage2nix.yaml#L2475-L2536

like

```yaml
- haskell-lsp == 0.18.0.0 # for ghcide-0.4 (or whatever is the correct ghcide version)
```

Once this gets merge into `haskell-updates` and the update script run, it will create a `haskell-lsp_0_18_0_0` derivation.  This will exist as long as the `haskell-lsp == 0.18.0.0` line exists.

You should be able to use this package in an override to `ghcide`.  That will hopefully get ghcide building.

**Comment by turion** [comment]
> @turion You are discovering why people use `haskell.nix`.

Hehe, yes ;) but I haven't given up on `nixpkgs` becoming good enough for standard usage yet.

**Comment by turion** [comment]
@cdepillabout I guess I could, but this is quite a manual and tiresome cycle. I'll just wait until upstream has released a new version. https://github.com/digital-asset/ghcide/issues/277

This whole workflow feels like it's going to break every time e.g. `haskell-lsp` releases a newer version than `ghcide` can handle. Is that correct? How would this situation ever stabilize?

**Comment by cdepillabout** [comment]
> This whole workflow feels like it's going to break every time e.g. haskell-lsp releases a newer version than ghcide can handle. Is that correct?

For the most part, yes.

> How would this situation ever stabilize?

There are some cases where the situation would stabilize:

-   On `release` branches of nixpkgs.

    `release` branches of `nixpkgs` don't get updates to the Haskell package versions.  So if you get `ghcide` working on one of the `release` branches, it will probably stay working forever.

-   If `ghcide` and `haskell-lsp` are released to stackage (and end up in an LTS release).

    nixpkgs tracks the latest LTS release, so packages from the LTS releases are almost always working, because they are all "known compatible".

    If the upstream authors of `ghcide` and `haskell-lsp` don't want to get compatible versions into stackage, maybe you could?

-   If you add specific versions of `haskell-lsp` and `ghcide` to `extra-packages` in `configuration-hackage2nix.nix` as suggested above.

    This will generate versions of `haskell-lsp` and `ghcide` like `haskell-lsp_0_18_0_0` and `ghcide_0_4_0_0` (or whatever) that you can use in overrides and are known to work together.

    These should continue working at least until other dependencies get too new.

-    If `ghcide` and `haskell-lsp` stop frequently releasing new versions to Hackage.

-------------------------------------------------------------------

Unfortunately, the Haskell stuff in nixpkgs doesn't work too well for new packages that aren't in Stackage and have frequent releases to Hackage.

As suggested by mpickering, creating an overlay with known package versions (or using a tool like `haskell.nix` to do this for you) might be the easiest solution.

Creating a `ghcide.nix` github repo that other people could also use might be helpful for the community as well.

**Comment by turion** [comment]
@cdepillabout Thank again you for your detailed explanations! It certainly helps me to understand the process and to more effectively contribute.

> If ghcide and haskell-lsp are released to stackage (and end up in an LTS release).

Ah, good point! So LTS in Haskell on `nixpkgs` always refers to `stackage` LTS, I'm assuming now. There is a ticket tracking that: https://github.com/digital-asset/ghcide/issues/144
This sounds like the best option for me then. I'll see whether I can help with that.

**Comment by peti** [comment]
I have attempted to fix the `ghcide` build in Nixpkgs, too, but it is quite a mess. Honestly, I think it's much easier to work with upstream to fix `ghcide`'s build with recent versions of its dependencies than trying to build it in Nixpkgs with old ones.

**Comment by cdepillabout** [comment]
I think @mdorman might have gotten ghcide working in https://github.com/NixOS/nixpkgs/pull/76103.

**Comment by domenkozar** [comment]
This is why I maintain https://github.com/hercules-ci/ghcide-nix

**Comment by peti** [comment]
@turion, do you still intend to update this PR? Or can it be closed?

**Pull request closed without merging**
